core/src/main/java/com/linecorp/armeria/server/http/dynamic/DynamicPath.java

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright 2016 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server.http.dynamic;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.StringJoiner;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+
+/**
+ * A bindable path. It holds three things:
+ *
+ * <ul>
+ *   <li> A regex-compiled form of the path. It is used for matching and binding.
+ *   <li> A skeleton of the path. It is used for duplication detecting.
+ *   <li> A set of variables declared in the path.
+ * </ul>
+ *
+ * @see DynamicHttpService
+ * @see RequestMapping
+ * */
+public final class DynamicPath {
+    private static final Pattern VALID_PATTERN = Pattern.compile("(/[^/]+)+");
+
+    /**
+     * Create a {@link DynamicPath} instance from given str.
+     * */
+    public static DynamicPath of(String str) {
+        if (!VALID_PATTERN.matcher(str).matches()) {
+            throw new IllegalArgumentException(
+                    String.format("%s: Invalid DynamicPath (valid: %s)", str, VALID_PATTERN));
+        }
+        return new DynamicPath(str);
+    }
+
+    /**
+     * Regex form of given path, which will be used for matching or binding.
+     *
+     * <p>
+     * Ex) "/{x}/{y}/{x}" -> "/(?&lt;x&gt;[^/]+)/(?&lt;y&gt;[^/]+)/(\\k&lt;x&gt;)"
+     * </p>
+     * */
+    private final Pattern pattern;
+
+    /**
+     * Skeletal form of given path, which is used for duplicated routing rule detection.
+     * For example, "/{a}/{b}" and "/{c}/{d}" has same skeletal form and regarded as duplicated.
+     *
+     * <p>
+     * Ex) "/{x}/{y}/{z}" -> "/{}/{}/{}"
+     * </p>
+     * */
+    private final String skeleton;
+
+    /**
+     * Set of the variables declared in this dynamic path.
+     * */
+    private final ImmutableSet<String> variables;
+
+    DynamicPath(String str) {
+        StringJoiner patternJoiner = new StringJoiner("/");
+        StringJoiner skeletonJoiner = new StringJoiner("/");
+        Set<String> placeholders = Sets.newHashSet();
+        for (String token : str.split("/")) {
+            if (token.startsWith("{") && token.endsWith("}")) {
```

Comment:
Could we also support the pattern of `:var`? e.g. `/:x/:y/:z`? (because it's simpler and as popular as `{...}`)